x-pack/plugins/ingest_pipelines/public/application/components/pipeline_editor/components/processor_form/processors/reroute.tsx

```diff
@@ -0,0 +1,126 @@
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import React, { FunctionComponent, useEffect, useState } from 'react';
+import { i18n } from '@kbn/i18n';
+import { FormattedMessage } from '@kbn/i18n-react';
+import { EuiCode } from '@elastic/eui';
+
+import { get } from 'lodash';
+import {
+  ComboBoxField,
+  FIELD_TYPES,
+  UseField,
+  Field,
+  useFormContext,
+} from '../../../../../../shared_imports';
+
+import { FieldsConfig, to, from } from './shared';
+
+const fieldsConfig: FieldsConfig = {
+  /* Optional field configs */
+  destination: {
+    type: FIELD_TYPES.TEXT,
+    serializer: from.emptyStringToUndefined,
+    label: i18n.translate('xpack.ingestPipelines.pipelineEditor.reroute.destinationFieldLabel', {
+      defaultMessage: 'Destination (optional)',
+    }),
+    helpText: (
+      <FormattedMessage
+        id="xpack.ingestPipelines.pipelineEditor.reroute.destinationFieldHelperText"
+        defaultMessage="A static value for the target. Cannot be set when {dataset} or {namespace} is set."
+        values={{
+          dataset: <EuiCode>{'dataset'}</EuiCode>,
+          namespace: <EuiCode>{'namespace'}</EuiCode>,
+        }}
+      />
+    ),
+  },
+  dataset: {
+    defaultValue: null,
+    type: FIELD_TYPES.COMBO_BOX,
+    deserializer: to.arrayOfStrings,
+    serializer: from.optionalArrayOfStrings,
+    label: i18n.translate('xpack.ingestPipelines.pipelineEditor.reroute.datasetFieldLabel', {
+      defaultMessage: 'Dataset (optional)',
+    }),
+    helpText: (
+      <FormattedMessage
+        id="xpack.ingestPipelines.pipelineEditor.reroute.datasetFieldHelperText"
+        defaultMessage="Field references or a static value for the dataset part of the data stream name. In addition to the criteria for index names, cannot contain {dash} and must be no longer than 100 characters. Defaults to {defaultValue}."
+        values={{
+          dash: <EuiCode>{'-'}</EuiCode>,
+          defaultValue: <EuiCode>{'{{data_stream.dataset}}'}</EuiCode>,
+        }}
+      />
+    ),
+  },
+  namespace: {
+    defaultValue: null,
+    type: FIELD_TYPES.COMBO_BOX,
+    deserializer: to.arrayOfStrings,
+    serializer: from.optionalArrayOfStrings,
+    label: i18n.translate('xpack.ingestPipelines.pipelineEditor.reroute.namespaceFieldLabel', {
+      defaultMessage: 'Namespace (optional)',
+    }),
+    helpText: (
+      <FormattedMessage
+        id="xpack.ingestPipelines.pipelineEditor.reroute.namespaceFieldHelperText"
+        defaultMessage="Field references or a static value for the namespace part of the data stream name. Should meet the criteria for index names and must be no longer than 100 characters. Defaults to {defaultValue}."
+        values={{ defaultValue: <EuiCode>{'{{data_stream.namespace}}'}</EuiCode> }}
+      />
+    ),
+  },
+};
+
+export const Reroute: FunctionComponent = () => {
+  const form = useFormContext();
```

Comment:
I think we might be able to simplify this by using the `useFormData` hook:

```typescript
const [{ fields }] = useFormData({ watch: ['fields.dataset', 'fields.namespace'] });
...

   <UseField
        data-test-subj="destinationField"
        config={fieldsConfig.destination}
        component={Field}
        componentProps={{
          euiFieldProps: {
            disabled: !!fields?.dataset || !!fields?.namespace,
          },
        }}
        path="fields.destination"
      />

```

Reply:
I like this, thanks for the suggestion! I just changed it a little bit so that the Destination value gets cleared up if the Dataset or the Namespace is set.

Reply:
Edit: actually with the new logic, the Destination parameter doesn't really get cleared up (I don't know why I thought it does), which is why one of the tests failed. I also tried setting the `value` parameter in the `euiFieldProps` instead, but this didn't work either:
```
      <UseField
        data-test-subj="destinationField"
        config={fieldsConfig.destination}
        component={Field}
        componentProps={{
          euiFieldProps: {
            disabled: fields?.dataset.length > 0 || fields?.dataset.length > 0,
            value:
              fields?.dataset.length > 0 || fields?.dataset.length > 0 ? '' : fields?.destination,
          },
        }}
        path="fields.destination"
      />
```

So I guess I should return the previous logic, unless you have another suggestion?

Reply:
I think that happens because the component is self managed, perhaps a small useEffect can be used to set the value with the use form hook:

```typescript
      const { form } = useForm();
      const [{ fields }] = useFormData({ watch: ['fields.dataset', 'fields.namespace'] });
    
      useEffect(() => {
        if (fields?.dataset.length > 0 || fields?.namespace.length > 0) {
          form.setFieldValue('fields.destination', undefined);
        }
      }, [fields]);

      <UseField
        ...
        componentProps={{
          euiFieldProps: {
            disabled: fields?.dataset.length > 0 || fields?.namespace.length > 0,
          },
        }}
        path="fields.destination"
      />
```

Perhaps something like that?

Reply:
Thanks for the suggestion @sabarasaba, this logic works well for clearing up the Destination field!

I only made one small change: instead of setting the Destination value to `undefined`, I set it to an empty string, because setting it to `undefined` doesn't clear up the text in the field and if the user re-enables this field, they may have the impression that it is already filled in while it's not:

https://github.com/elastic/kibana/assets/59341489/c8c24022-f11d-419b-883b-fcaf8f6d222e